src/hotspot/share/opto/vectorization.hpp

```diff
@@ -48,15 +48,128 @@ class VTrace : public StackObj {
+    _phase     (lpt->_phase),
+    _lpt       (lpt),
+    _allow_cfg (allow_cfg),
+    _cl        (nullptr),
+    _cl_exit   (nullptr),
+    _iv        (nullptr) {}
+  NONCOPYABLE(VLoop);
+
+  IdealLoopTree* lpt()        const { return _lpt; };
+  PhaseIdealLoop* phase()     const { return _phase; }
+  CountedLoopNode* cl()       const { return _cl; };
+  Node* cl_exit()             const { return _cl_exit; };
+  PhiNode* iv()               const { return _iv; };
+  int iv_stride()             const { return cl()->stride_con(); };
+  bool is_allow_cfg()         const { return _allow_cfg; }
+
+  CountedLoopEndNode* pre_loop_end() const {
+    assert(cl()->is_main_loop(), "only main loop can reference pre-loop");
+    assert(_pre_loop_end != nullptr, "must have found it");
+    return _pre_loop_end;
+  };
+
+  CountedLoopNode* pre_loop_head() const {
+    CountedLoopNode* head = pre_loop_end()->loopnode();
+    assert(head != nullptr, "must find head");
+    return head;
+  };
+
+  // Estimate maximum size for data structures, to avoid repeated reallocation
+  int estimated_body_length() const { return lpt()->_body.size(); };
+  int estimated_node_count()  const { return (int)(1.10 * phase()->C->unique()); };
+
+#ifndef PRODUCT
+  const VTrace& vtrace()      const { return _vtrace; }
+
+  bool is_trace_preconditions() const {
+    return vtrace().is_trace(TraceAutoVectorizationTag::PRECONDITIONS);
+  }
+
+  bool is_trace_pointer_analysis() const {
+    return vtrace().is_trace(TraceAutoVectorizationTag::POINTER_ANALYSIS);
+  }
+#endif
+
+  // Is the node in the basic block of the loop?
+  // We only accept any nodes which have the loop head as their ctrl.
+  bool in_bb(const Node* n) const {
+    const Node* ctrl = _phase->has_ctrl(n) ? _phase->get_ctrl(n) : n;
+    return n != nullptr && n->outcnt() > 0 && ctrl == _cl;
+  }
+
+  // Check if the loop passes some basic preconditions for vectorization.
+  // Return indicates if analysis succeeded.
+  bool check_preconditions();
+
+private:
+  const char* check_preconditions_helper();
+};
+
+// Optimization to keep allocation of large arrays in AutoVectorization low.
+// We allocate the arrays once, and reuse them for multiple loops that we
+// AutoVectorize, clearing them before every new use.
+class VSharedData : public StackObj {
+private:
+  // Arena, used to allocate all arrays from.
+  Arena _arena;
+
+  // An array that maps node->_idx to a much smaller idx, which is at most the
+  // size of a loop body. This allow us to have smaller arrays for other data
+  // structures, since we are using smaller indices.
+  GrowableArray<int> _node_idx_to_loop_body_idx;
+
+public:
+  VSharedData() :
+    _arena(mtCompiler),
+    _node_idx_to_loop_body_idx(&_arena, estimated_node_count(), 0, 0)
+  {
+  }
+
+  GrowableArray<int>& node_idx_to_loop_body_idx() {
+    // Since this is a shared resource, we clear before every individual use.
+    _node_idx_to_loop_body_idx.clear();
```

Comment:
I think it should be explicit `VSharedData::clear()` method called in `auto_vectorize()`. Otherwise much later someone will have hard time to find place where the space is cleared.

Reply:
@vnkozlov done.